pkg/util/util.go

```diff
@@ -15,135 +15,15 @@ package util
-			},
-		}
-		terms = append(terms, term)
-	}
-	return &corev1.PodAntiAffinity{
-		PreferredDuringSchedulingIgnoredDuringExecution: terms,
-	}
-}
-
-// AffinityForNodeSelector creates an Affinity for NodeSelector
-// Externally we use NodeSelector for simplicity,
-// while internally we convert it to affinity which can express complex scheduling rules
-func AffinityForNodeSelector(namespace string, required bool, antiLabels, selector map[string]string) *corev1.Affinity {
-	if selector == nil {
-		return nil
-	}
-	affinity := &corev1.Affinity{}
-	if antiLabels != nil {
-		affinity.PodAntiAffinity = AntiAffinityForPod(namespace, antiLabels)
-	}
-
-	keys := []string{}
-	for key := range selector {
-		keys = append(keys, key)
-	}
-	sort.Strings(keys) // we must use sorted selector, otherwise affinity may vary causing new statefulset generated and pod recreated
-
-	requiredTerms := []corev1.NodeSelectorTerm{}
-	if required { // all nodeSelectors are required
-		var exps []corev1.NodeSelectorRequirement
-		for _, key := range keys {
-			requirement := corev1.NodeSelectorRequirement{
-				Key:      key,
-				Operator: corev1.NodeSelectorOpIn,
-				Values:   strings.Split(selector[key], ","),
-			}
-			// NodeSelectorRequirement in the same MatchExpressions are ANDed otherwise ORed
-			exps = append(exps, requirement)
-		}
-		requiredTerms = append(requiredTerms, corev1.NodeSelectorTerm{MatchExpressions: exps})
-		affinity.NodeAffinity = &corev1.NodeAffinity{
-			RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
-				NodeSelectorTerms: requiredTerms,
-			},
-		}
-		return affinity
-	}
-
-	preferredTerms := []corev1.PreferredSchedulingTerm{}
-	exps := []corev1.NodeSelectorRequirement{}
-	for _, key := range keys {
-		if selector[key] == "" {
-			continue
-		}
-		values := strings.Split(selector[key], ",")
-		// region,zone,rack,host are preferred labels, others are must match labels
-		if weight, ok := topologySchedulingWeight[key]; ok {
-			t := corev1.PreferredSchedulingTerm{
-				Weight: weight,
-				Preference: corev1.NodeSelectorTerm{
-					MatchExpressions: []corev1.NodeSelectorRequirement{
-						{
-							Key:      key,
-							Operator: corev1.NodeSelectorOpIn,
-							Values:   values,
-						},
-					},
-				},
-			}
-			preferredTerms = append(preferredTerms, t)
-		} else {
-			requirement := corev1.NodeSelectorRequirement{
-				Key:      key,
-				Operator: corev1.NodeSelectorOpIn,
-				Values:   values,
-			}
-			// NodeSelectorRequirement in the same MatchExpressions are ANDed otherwise ORed
-			exps = append(exps, requirement)
-		}
-	}
-	requiredTerms = append(requiredTerms, corev1.NodeSelectorTerm{MatchExpressions: exps})
-
-	affinity.NodeAffinity = &corev1.NodeAffinity{
-		RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
-			NodeSelectorTerms: requiredTerms,
-		},
-		PreferredDuringSchedulingIgnoredDuringExecution: preferredTerms,
-	}
-
-	return affinity
-}
-
```

Comment:
If I understand correctly, `nodeSelector` equals to pod.spec.nodeSelector when `nodeselectorRequired` is true?

How about remove `nodeselectorRequired` only, but keep `nodeSelector` . It can be copied directly to pod spec, I guess. Then the user can configure scheduling policies of TiDB, PD, TiKV like normal pods.

Reply:
vote for it

Reply:
oh, there is a difference, our `nodeSelector` accepts comma-separated values which is not supported by `pod.spec.nodeSelector`

Reply:
we can change our `nodeSelector` to be the same as `pod.spec.nodeSelector`?

Reply:
however, instead of removing it, I think it's better to keep it even if we need to drop the comma-separated values support
comma-separated values support is not necessary because we are able to configure in affinity

Reply:
The previous`nodeSelector` was self-defined, and it was converted to `affinity` in the background.
the suggestion is nice.

Reply:
@tennix  have any opinion?